### PR TITLE
array_merge lost non-empty-string keys type

### PR DESCRIPTION
**Pull request by staabm**
suggestion for a fix of bug #6927. 

I am wondering why the `generalize` call was in the `ArrayMergeFunctionDynamicReturnTypeExtension` 

.. the only thing I could think of are "cosmetic" reasons for error messages, since the types we get when dropping the `generalize` call seem to be more precise and still valid? lets see what CI thinks about it.

the `generalize` call was added a long time ago with https://github.com/phpstan/phpstan-src/commit/002aa73b8e5614aca9fda55dddb84e82007ef93e 

closes https://github.com/phpstan/phpstan/issues/6927

### CODE REVIEW

**Comment by staabm** [comment]
unrelated: there is a interessting test-case, I tried to add.

```
	/**
	 * @param array<numeric-string&non-empty-string, string> $params1
	 * @param array<numeric-string, string> $params2
	 */
	function foo5(array $params1, array $params2): void
	{
		$params2 = array_merge($params1, $params2);

		assertType('array<numeric-string, string>', $params2);
	}
``` 

it errors with 
```
There was 1 failure:

1) PHPStan\Analyser\NodeScopeResolverTest::testFileAsserts with data set "C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser/data/bug-6927.php:61" ('type', 'C:\dvl\Workspace\phpstan-src-...27.php', PHPStan\Type\Constant\ConstantStringType Object (...), PHPStan\Type\ArrayType Object (...), 61)
Expected type array<numeric-string, string>, got type array<int, string> in C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser/data/bug-6927.php on line 61.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'array<numeric-string, string>'
+'array<int, string>'

C:\dvl\Workspace\phpstan-src-staabm\src\Testing\TypeInferenceTestCase.php:97
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\NodeScopeResolverTest.php:846
```

which is a different bug IMO.. but interessting nevertheless

**edit**: [opened a new issue for that](https://github.com/phpstan/phpstan/issues/6945)

**Comment by ondrejmirtes** [comment]
I agree, it shouldn't be generalized.

**Comment by staabm** [comment]
the remaining error

``` 
Error: Property Composer\Util\Http\CurlDownloader::$jobs (
array<array{url: string, origin: string, attributes: array{retryAuthFailure: bool, redirects: int, retries: int, storeAuth: bool}, options: array, progress: array, curlHandle: resource, filename: string|false, headerHandle: resource, ...}>) 
does not accept 
non-empty-array<array{url: string, origin: string, attributes: array{retryAuthFailure: bool, redirects: int, retries: int, storeAuth: bool}, options: array, progress: array, curlHandle: CurlHandle|resource, filename: string|false|null, headerHandle: resource, ...}>.
```

seems legit and unrelated, because `resource` and `CurlHandle|resource`, and also `string|false` and `string|false|null` do not match.
see https://github.com/composer/composer/blob/3ae111140facdba8ae82adcd1085e4adfc7d715c/src/Composer/Util/Http/CurlDownloader.php#L261-L273

the code seems not to be dependent on `array_merge` which is the one we change with this PR.

File: tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php [hunk 1]
```diff
@@ -4707,31 +4707,31 @@ public function dataArrayFunctions(): array
 				'array_values($generalStringKeys)',
 			],
 			[
-				'non-empty-array<int|(literal-string&non-empty-string), stdClass>',
+				'array{foo: stdClass, 1: stdClass}',
```
Comment (staabm) [inline tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php:4710]: changed existing tests, to more specific types - as the PR gets better results in these cases.

File: tests/PHPStan/Analyser/data/bug-6927.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	function foo4(array $params1, array $params2): void
+	{
+		$params2 = array_merge($params1, $params2);
+
+		assertType('array<non-empty-string, string>', $params2);
+	}
+
+	/**
+	 * @param array{return: int, stdout: string, stderr: string} $params1
+	 * @param array{return: int, stdout?: string, stderr?: string} $params2
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/bug-6927.php:55]: this test is inspired by a scenario which failed with a previous version of this PR within the composer code-base

File: src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php [hunk 3]
```diff
@@ -30,6 +32,41 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
 		}
 
+		$allConstant = true;
+		foreach ($functionCall->getArgs() as $arg) {
+			$argType = $scope->getType($arg->value);
+			if ($arg->unpack || !$argType instanceof ConstantArrayType) {
+				$allConstant = false;
```
Comment (ondrejmirtes) [inline src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php]: I like the result of this PR but the implementation could be improved - you're iterating over `->getArgs()` twice and you're getting the `$arg->value` type twice. Instead, you could save the argument types to an array and the second foreach could iterate over that.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php [hunk 4]
```diff
@@ -26,16 +28,54 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
...
+				break;
+			}
+		}
+
+		if ($allConstant) {
+			$newArrayBuilder = ConstantArrayTypeBuilder::createEmpty();
+			foreach ($argTypes as $argType) {
+				if (!$argType instanceof ConstantArrayType) {
+					throw new ShouldNotHappenException();
```
Comment (ondrejmirtes) [inline src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php:56]: You could push only ConstantArrayType into `$argTypes` and this condition would not be necessary :)

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php (hunk 4 above)
Comment (staabm) [reply src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php:56]: but then I couldn't re-use the `$artTypes` for the 2nd-foreach loop..?

or I have all non-constant-array-args to be resolved twice via `scope->getType`
or alternatively I need 2 arrays to hold the different args per type

**Comment by ondrejmirtes** [comment]
Thank you, let's try this!

**Comment by ondrejmirtes** [comment]
The same thing that happened last time someone tried to do this happened again :( https://github.com/phpstan/phpstan-src/pull/641#issuecomment-908133888

This version is about 30 % slower on Slevomat's codebase than previous commit. I'm gonna revert this. Can you please send a new PR that just removes the generalization to see if it fixes the related issues? https://github.com/phpstan/phpstan-src/pull/1135/files#diff-235402991c12d4f1091a88a9c19ed82c38749ed4e4dce1322474ed6c6d7fdcc8L47-R92

**Comment by ondrejmirtes** [comment]
Oh, previous commit is also slow, I overlooked it. Well, we have a rabbit hole to go through.

**Comment by ondrejmirtes** [comment]
Alright, it's the same speed as the last stable version, we can release this as it is. (It might have something to do I'm on battery right now, I'm used to seeing faster numbers though.)

**Pull request merged**